### ISSUE TITLE
prune_heads() method for AutoModelForCausalLM

### ISSUE THREAD

**Issue by mnauf** [huggingface/transformers#33225]
### Feature request

We have `prune_heads()` method for `AutoModel` class, but not for `AutoModelForCausalLM`. Please provide `prune_heads()` method to `AutoModelForCausalLM` class.

### Motivation

Mechanistic interpretability study from the paper [Are Sixteen Heads Really Better than One?](https://arxiv.org/abs/1905.10650) explore how pruning attention heads effect the model's (BERT) performance. I want to do something similar for LLMs. Since LLMs are supported via `AutoModelForCausalLM` class, I am unable to do this experiment.

### Your contribution

Just getting started with HuggingFace. Not confident about implementing this feature on my own as of now.

**Comment by LysandreJik** [comment]
Hey @mnauf! Over the years we've seen very limited interest in this feature, hence why it wasn't implemented in recent LLMs.

If we see interest for it growing then we might consider it; the best way to approach it might be for you to do a fork with your changes to kickstart your experiments.

**Comment by shwu-nyunai** [comment]
@LysandreJik if not for a prune_heads() method, do you think that, with the recent explorations around llm pruning, we could at least have a structure in place for loading a pruned model? Ideally, depth/width pruned models be loadable. Just how we do it for quantized models, having a support in place for pruned models would be useful. 
I'd love to be pointed towards an already open/closed discussion/thread on the same. 
Thanks.

**Comment by ArthurZucker** [comment]
Yeah having something related to pruning does not sound too bad! You can already load a model with less width, with `ignore_missmatched_sizes = True` in `from_pretrained`. For width, you do need to update the config. 

On the hub we could do stuff tho to link pruned versions !

**Comment by qgallouedec** [comment]
Hey @mnauf, sorry for the slow reply, we're catching up on the backlog.

Rather than extending it to causal LMs, head pruning (and head masking) was removed from the library altogether in v5, since almost no recent model supported it. Check the migration guide.

https://github.com/huggingface/transformers/blob/624b32a2c5067cd6df2c140e6a7371cc9464f05d/MIGRATION_GUIDE_V5.md#L434-L437

For head-ablation experiments on LLMs, a forward pre-hook on each layer's `self_attn.o_proj` that zeroes the `head_dim`-wide slice of its input for the heads you want to drop gives you the same effect without touching the model code.
